ssl/ssl_quic.c

```diff
@@ -0,0 +1,250 @@
+        ssl->quic_input_data_tail = qd;
+
+        data += l;
+        len -= l;
+    }
+
+    return 1;
+}
+
+int SSL_CTX_set_quic_method(SSL_CTX *ctx, const SSL_QUIC_METHOD *quic_method)
+{
+    switch (ctx->method->version) {
+    case DTLS1_VERSION:
+    case DTLS1_2_VERSION:
+    case DTLS_ANY_VERSION:
+    case DTLS1_BAD_VER:
+        return 0;
+    default:
+        break;
+    }
+    ctx->quic_method = quic_method;
+    ctx->options &= SSL_OP_ENABLE_MIDDLEBOX_COMPAT;
+    return 1;
+}
+
+int SSL_set_quic_method(SSL *ssl, const SSL_QUIC_METHOD *quic_method)
+{
+    switch (ssl->method->version) {
+    case DTLS1_VERSION:
+    case DTLS1_2_VERSION:
+    case DTLS_ANY_VERSION:
+    case DTLS1_BAD_VER:
+        return 0;
+    default:
+        break;
+    }
+    ssl->quic_method = quic_method;
+    ssl->options &= SSL_OP_ENABLE_MIDDLEBOX_COMPAT;
+    return 1;
+}
+
+int quic_set_encryption_secrets(SSL *ssl, OSSL_ENCRYPTION_LEVEL level)
+{
+    uint8_t *read_secret = NULL;
+    uint8_t *write_secret = NULL;
+    static const unsigned char zeros[EVP_MAX_MD_SIZE];
+
+    if (!SSL_IS_QUIC(ssl))
+        return 1;
+
+    /* secrets from the POV of the client */
+    switch (level) {
+    case ssl_encryption_early_data:
+        write_secret = ssl->early_secret;
+        break;
+    case ssl_encryption_handshake:
+        read_secret = ssl->client_finished_secret;
+        write_secret = ssl->server_finished_secret;
+        break;
+    case ssl_encryption_application:
+        read_secret = ssl->client_app_traffic_secret;
+        write_secret = ssl->server_app_traffic_secret;
+        break;
+    default:
+        return 1;
+    }
+    /* In some cases, we want to set the secret only when BOTH are non-zero */
+    if (read_secret != NULL && write_secret != NULL
+            && !memcmp(read_secret, zeros, ssl->quic_len)
+            && !memcmp(write_secret, zeros, ssl->quic_len))
+        return 1;
+
+    if (ssl->server) {
+        if (!ssl->quic_method->set_encryption_secrets(ssl, level, read_secret,
+                                                      write_secret, ssl->quic_len)) {
+            SSLfatal(ssl, SSL_AD_INTERNAL_ERROR, SSL_F_QUIC_SET_ENCRYPTION_SECRETS,
+                     ERR_R_INTERNAL_ERROR);
+            return 0;
+        }
+    } else {
+        if (!ssl->quic_method->set_encryption_secrets(ssl, level, write_secret,
+                                                      read_secret, ssl->quic_len)) {
+            SSLfatal(ssl, SSL_AD_INTERNAL_ERROR, SSL_F_QUIC_SET_ENCRYPTION_SECRETS,
+                     ERR_R_INTERNAL_ERROR);
+            return 0;
+        }
+    }
+
+    return 1;
+}
+
+int SSL_process_quic_post_handshake(SSL *ssl)
+{
+    if (SSL_in_init(ssl) || !SSL_IS_QUIC(ssl)) {
+        SSLerr(SSL_F_SSL_PROCESS_QUIC_POST_HANDSHAKE, ERR_R_SHOULD_NOT_HAVE_BEEN_CALLED);
+        return 0;
+    }
+
+    ossl_statem_set_in_init(ssl, 1);
+
+    if (ssl->handshake_func(ssl) <= 0)
+        return 0;
```

Comment:
It looks like SSL_in_init() returns 0 again after the first call of this function.  That means calling this function twice results in "you should not call this function" error.  If I understand it correctly, this function can be called multiple times after handshake completes.  Session ticket can be received in any time.
I think we need to call ossl_statem_set_in_init(ssl, 0) after ssl->handshake_func regardless of its return value (or when it returns SSL_ERR_WANT_READ).

Reply:
The test code calls it multiple times to handle the default number of NewSessionTicket messages (2) (and they succeed).
The `handshake_func()` ends up in the TLS_ST_OK state, after handling the NewSessionTicket, which does an `ossl_statem_set_in_init(ssl, 0)` via `tls_finish_handshake()`. So, the extra call to `ossl_statem_set_in_init()` should not be necessary.

Reply:
While I was testing with **real** quic client, ssl->handshake_func(ssl) returned 0 and indicated SSL_ERR_WANT_READ.  SSL_in_init(ssl) returned 1 then.  Next call of SSL_process_quic_post_handshake failed with "you should not call" error.
I think we need to call ossl_statem_set_in_init(ssl, 0) after handshake_func.
I think that it should be allowed to call SSL_process_quic_post_handshake() without providing any data at all.